### ISSUE TITLE
cmd/compile: panics with generic type alias and based-on constraint

### ISSUE THREAD

**Issue by splace** [golang/go#68580]
### Go version

1.23rc1

### Output of `go env` in your module/workspace:

```shell
n/a
```

### What did you do?

https://go.dev/play/p/kMvT8_f91aK?v=gotip

### What did you see happen?

<unknown line number>: internal compiler error: panic: /usr/local/go/src/cmd/compile/internal/types2/subst.go:56: assertion failed

etc.

### What did you expect to see?

no panic

**Comment by ianlancetaylor** [comment]
CC @griesemer @timothy-king

**Comment by griesemer** [comment]
Reproduced at 1.23rc2. Interestingly, the code typechecks fine when checked by the typechecker alone.

**Comment by timothy-king** [comment]
Reduced example: https://go.dev/play/p/X9GxAbUdgeA?v=gotip

**Comment by griesemer** [comment]
Simpler reproducer:
```Go
package main

type A[P any] = struct{ _ P }

type N[P any] A[P]

func f[P any](N[P]) {}

var _ = f[int]
```

**Comment by cuonglm** [comment]
@griesemer Do we need to backport this to 1.23?

**Comment by cuonglm** [comment]
@gopherbot please backport this issue to go1.23

**Comment by griesemer** [comment]
@cuonglm Yes, we should backport this. I believe an issue is not needed per @dmitshur 's e-mail:

> Any CLs that land after [CL 600176](https://go.dev/cl/600176) ("internal/goversion: update Version to 1.24") but target Go 1.23 must be cherry-picked to release-branch.go1.23. Cherry-pick CLs can be created following [the same steps](https://go.dev/wiki/MinorReleases#making-cherry-pick-cls) as usual but do not need a backport issue or approval from the release team. Please continue to use [your judgment](https://go.dev/s/release#june--december-week-2-release-candidate-1-issued) about what should be included this late in the Go 1.23 release cycle.
